Use hugepages on hpe-cray-ex systems.

jhh67 opened the pull request:
Use hugepages on hpe-cray-ex systems.

Resolves Cray/chapel-private#3141 and closes https://github.com/Cray/chapel-private/issues/4844.

Signed-off-by: John H. Hartman <jhh67@users.noreply.github.com>

ronawho commented:
In my initial performance experiments, I'm seeing worse performance with ISx and indexgather:

ISx:
```
chpl test/studies/isx/isx-hand-optimized.chpl --fast
./isx-hand-optimized -nl 16 --mode=weakISO --useSubTimers=true --n=$((2**26))

# no hugepages
total = 3.95756 (3.88047..4.01847)

# 16M hugepages
total = 5.97763 (5.77893..6.30889)
```

indexgather:
```
chpl test/studies/bale/aggregation/ig.chpl --fast -suseBlockArr --no-optimize-forall-unordered-ops
./ig -nl 16 --N=$((2**25))

# no hugepages
       AGG:	6.149 seconds	11.176 GB/s/node

# 16M hugepages 
       AGG:	9.107 seconds	7.546 GB/s/node
```

I think we need more time to investigate the performance loss, so let's hold off on this for the 1.26 release and come back to it afterwards.

ronawho commented:
Let's convert this to a draft or close based on the current lackluster performance results

jlbyrne-hpe commented:
When testing this with the 1.30.0 source, I needed to make the following additional change to get it to link:
```
diff --git a/util/chplenv/compile_link_args_utils.py b/util/chplenv/compile_link_args_utils.py
index 89894947c4..214fad261f 100644
--- a/util/chplenv/compile_link_args_utils.py
+++ b/util/chplenv/compile_link_args_utils.py
@@ -151,12 +151,11 @@ def compute_internal_compile_link_args(runtime_subdir):
     elif chpl_comm.get() == 'gasnet':
         extend2(tgt_compile, chpl_gasnet.get_compile_args())
         extend2(tgt_link, chpl_gasnet.get_link_args())
-    elif chpl_comm.get() == 'ugni':
-        # If there isn't a hugepage module loaded, we need to request
-        # libhugetlbfs ourselves.
-        pe_product_list = os.environ.get('PE_PRODUCT_LIST', None)
-        if pe_product_list and 'HUGETLB' in pe_product_list:
-            tgt_link[1].append('-lhugetlbfs')
+    # If there isn't a hugepage module loaded, we need to request
+    # libhugetlbfs ourselves.
+    pe_product_list = os.environ.get('PE_PRODUCT_LIST', None)
+    if pe_product_list and 'HUGETLB' in pe_product_list:
+        tgt_link[1].append('-lhugetlbfs')
 
     if chpl_tasks.get() == 'qthreads':
         extend2(tgt_compile, chpl_qthreads.get_compile_args())

```

ronawho commented:
Not sure how this didn't occur to me when we first ran into this, but this is probably because with hugepages memory is reserved/faulted-in at allocation by default. For dynamic registration under ugni we set `HUGETLB_NO_RESERVE` to avoid this fault-in (see https://github.com/chapel-lang/chapel/pull/6947). However, we don't set this for static registration because that would cause our fixed heap size determination to fail (see https://github.com/chapel-lang/chapel/pull/11242)

Testing with this, it does appear to be the source of the perf impact, at least for ig and 2M hugepages.


```
chpl test/studies/bale/aggregation/ig.chpl --fast -suseBlockArr --no-optimize-forall-unordered-ops

./ig-main -nl 16 --N=$((2**25))
       AGG:	6.033 seconds	11.390 GB/s/node
       
./ig-hp -nl 16 --N=$((2**25))
       AGG:	11.398 seconds	6.029 GB/s/node
       
HUGETLB_NO_RESERVE=yes ./ig-hp -nl 16 --N=$((2**25))
       AGG:	6.036 seconds	11.385 GB/s/node
```

ronawho commented:
If we can't use `HUGETLB_NO_RESERVE` I wonder if there's a way to request the hugepages get interleaved at fault-in time or if we could just use `mmap` directly with a hugepage flag.

jhh67 commented:
I can't think of any good solutions to this problem. I don't see a way to force the hugepages to be interleaved on a page fault, nor does there seem to be an `mmap` flag that would help. I'll keep looking, but if I don't find anything  we could

1. Create an environment variable that specifies the absolute heap size rather than a maximum. If this is set, we use it as the heap size rather than trying to determine the heap size. If the system doesn't have enough physical memory the process would likely die with an OOM error at runtime.
2. Fault-in the pages when sizing the heap, but subsequently free the heap and allocate a  new one without faulting-in the pages. I haven't tried this, so I'm not sure if it would even work or what the overhead might be.

jhh67 commented:
I implemented \# 2 above as an experiment and it does solve the performance problem that @ronawho reported. My "solution" introduces two issues, however. First, the `libhugetlbfs` library gets the value of the `HUGETLB_NO_RESERVE` environment variable once and stores it in a global variable. It isn't possible to change the value of this variable between calls to  `libhugetlbfs`. For the experiment I modified `get_huge_pages`, but this isn't feasible in production given our policy of not modifying third-party software. It might be possible to call `mmap` directly, but there is quite a lot of infrastructure around this call in the `libhugetlbfs` library that we would have to replicate.

Second, I have not measured the overhead of allocated a heap with physical pages mapped, freeing it, then allocating it again without physical pages mapped. It might be excessive.

jhh67 commented:
Closing this as this strategy won't work because when the heap is allocated there is no guarantee that the physical pages will be contiguous, so it may not be possible to allocate hugepages of the same size.
